### PR TITLE
feat: QnA import url

### LINKED ISSUE #3668: Composer should support lu file references
## Is your feature request related to a problem? Please describe.

The [.lu format](https://docs.microsoft.com/en-us/azure/bot-service/file-format/bot-builder-lu-file-format?view=azure-bot-service-4.0) used in composer support "exteneral references", which could be helpful in certain cases where user want to share there .lu content to other .lu files. But Composer don't support that, either referencing another .lu file or a certain section of a .lu file will result in error in Composer.
 
## Describe the solution you'd like

Support "external references"  in .lu files.  By implementing the resource resolver the lu parser needed to resolve those external references.

## Additional context
There might be additional work to support different kinds references, ie, whole file reference or section reference. In that case, whole file reference has a higher priority since it's the syntax that expected to be used in most cases.

### LINKED ISSUE #3872: Align QnA with the new design
The final design is here. https://www.figma.com/file/deFhMPBnRWxTYiYDYqYxrc/Composer-Vivek?node-id=7561%3A0

1. if expanded, no number needed: questions(number) => questions
2. color on selected and hovered should be different
3. auto-grow when typing question.
4.no (0), 0 is meaningless.
5. add QnA Pair, the new field should get auto focused
6. chevron icon and question space.
7. make the chevron icon in the middle. (left edge to question column)
8. spacing between two icons to both edge.
9. header top margin.
10. question and answer should have no hover and click state since nothing will do on that.
11. editing questions and answer should expand the row.
12. Something more on the vivke's design like how to delete the QA fields, show warning etc.

### LINKED ISSUE #3988: Fields flickering on language generation page
## Description
The fields on the Language Generation page flicker when moving focus from one field to another.

## Version
Latest build from main

## Browser

- [ ] Electron distribution
- [x] Chrome
- [ ] Safari
- [ ] Firefox
- [ ] Edge

## OS

- [X] macOS
- [ ] Windows
- [ ] Ubuntu

## To Reproduce

1. Open Language Generation page
2. Set focus on any field
3. Wait for a moment and then move the focus to another field

## Screenshots
Sep-02-2020 10-12-13

### PR DESCRIPTION
**Pull request by zhixzhan**
### QnA import

QnA import from url/scratch will create \<name\>.source.qna: 
```
> !# @source.url= <url>

...
```

For dialog.qna will import it by
```
[import](<name>.source.qna)
...
```

All up view can edit all imported qna file.  

### KB container based QnA structure.
Every created QnA container (from url/scratch) is a <name>.source.qna file, and shared by all dialog.qna, import by file id or file path both work, locale can be ignored.
```
{
  ...
  qna: 'knowledge-base/en-us/${BOTNAME}.en-us.qna',
  sourceQnA: 'knowledge-base/source/${FILENAME}.source.qna',
  dialogs: {
      qna: 'dialogs/${DIALOGNAME}/knowledge-base/en-us/${DIALOGNAME}.en-us.qna',
  }
}
```

##### refer by wildcard matcher \*/\**
- [x]  `[link name](<path>/*)` 
- [x]  `[link name](<path>/**)` 

##### refer a file
- [x]  `[link name](help.lu)` refer by id
- [x]  `[link name](help.en-us.lu)` when not specify locale will auto fallback to current luFile's locale 
- [x]  `[link name](./help.lu)` with path

##### refer a intent's utterances and patterns
- [x] `[link name](help.lu#HowToDrive)`
- [x] `[link name](help.lu#HowToDrive*utterances*)`
- [x] `[link name](help.lu#HowToDrive*patterns*)`

##### refer file's utterances and patterns
- [x] `[link name](help.lu#*utterances*)`
- [x] `[link name](help.lu#*patterns*)`
- [x] `[link name](help.lu#*utterancesAndPatterns*)`

### QnA reference
- [x]  `[link name](<path>/*)` 
- [x]  `[link name](<path>/**)` 
- [ ] `[link name](help.qna#$name?)`
- [ ] `[link name](help.qna#*alterations*?)`
- [ ] `[link name](help.qna#?question-to-find?)`
- [ ] `[link name](help.qna#*answers*?)`

## Known issue
1. Due to LU parser not support customized importResolver yet, client / LUEditor are not be able do intelligence auto-complete or find out imported file not existed. e.g. 
```
[link name](helpNotExist.lu)
```
only when publishing time  check error. this is tracked via https://github.com/microsoft/botframework-cli/issues/933

2. if a referenced intent not exist, the referenced' utterances will be empty,  and do not block publishing. e.g.  
```
[link name](help.lu#NotExistIntent)
```

## Task Item
close #3668 
close #3872 
close #3988

## Screenshots

image
image
image

### CODE REVIEW

**Comment by zhixzhan** [comment]
To use customize resolver, this PR is base on https://github.com/microsoft/BotFramework-Composer/pull/3778 doing qna import

**Comment by boydc2014** [comment]
A few things in technical side
1. let's use [import](faq.source.qna) instead of "[faq](faq.source.qna)". The latter approach also works, but brings more efforts to maintain. 
2. source.name is not needed in the options, because the file name indicates the source name.

File: Composer/packages/server/src/models/bot/botProject.ts [hunk 1]
```diff
@@ -346,7 +346,9 @@ export class BotProject implements IBotProject {
     }
     const file = this.files.get(name);
     if (file === undefined) {
-      throw new Error(`no such file ${name}`);
+      // throw new Error(`no such file ${name}`);
```
Comment (hatpick) [inline Composer/packages/server/src/models/bot/botProject.ts:349]: remove commented

File: Composer/packages/server/src/models/bot/builder.ts [hunk 2]
```diff
@@ -197,17 +199,19 @@ export class Builder {
     );
   }
 
-  private async runLuBuild(files: FileInfo[]) {
+  private async runLuBuild(files: FileInfo[], allFiles: FileInfo[]) {
     const config = await this._getConfig(files, 'lu');
     // if (config.models.length === 0) {
```
Comment (hatpick) [inline Composer/packages/server/src/models/bot/builder.ts:204]: remove commented

**Review by hatpick — CHANGES_REQUESTED** [review]

File: Composer/packages/client/src/components/EditableField.tsx [hunk 3]
```diff
@@ -1,61 +1,119 @@
...
   readonly?: boolean;
   required?: boolean;
   value?: string;
-
-  onChange: (newValue?: string) => void;
-  onFocus?: (id: string, value?: string) => void;
+  iconProps?: IconProps;
+  enableIcon?: boolean;
   onBlur?: (id: string, value?: string) => void;
+  onChange: (newValue?: string) => void;
```
[the remark this answers is not in the document]
Comment (zhixzhan) [reply Composer/packages/client/src/components/EditableField.tsx:61]: It's inherit from fluentUI TextField:
```onChange: (e, newValue?: string)```

File: Composer/packages/client/src/components/EditableField.tsx [hunk 4]
```diff
@@ -1,61 +1,119 @@
...
+  useEffect(() => {
+    if (componentFocusOnmount) {
+      fieldRef.current?.focus();
+    }
+  }, []);
   useEffect(() => {
     if (!hasBeenEdited || value !== localValue) {
       setLocalValue(value);
     }
   }, [value]);
```
[the remark this answers is not in the document]
Comment (zhixzhan) [reply Composer/packages/client/src/components/EditableField.tsx:104]: I guess not, here is sync external value change, internal change suppose to be not trigger this.

File: Composer/packages/client/src/recoilModel/types.ts [hunk 5]
```diff
@@ -140,10 +140,12 @@ export type BoilerplateVersion = {
...
 
-export enum QnAAllUpViewStatus {
-  Loading,
-  Success,
-  Failed,
-}
-
 export type Notification = CardProps & { id: string };
+
+export enum NotificationCardTypes {
```
[the remark this answers is not in the document]
Comment (zhixzhan) [reply Composer/packages/client/src/recoilModel/types.ts]: I thought the Notification types are enum, here set to enum type can strict created notifications to be supported types. @lei9444 do you think this change valid ?

[the remark this answers is not in the document]
Comment (zhixzhan) [reply Composer/packages/client/src/recoilModel/types.ts]: Offline talked to leilei, I see the diffrence

**Review by cwhitten — APPROVED** [review]

**Pull request merged**